src/inmanta/data/__init__.py

```diff
@@ -6614,16 +6667,72 @@ async def set_last_processed_model_version(
 CORE_SCHEMA_NAME = schema.CORE_SCHEMA_NAME
 
 
+def set_connection_pool(pool: asyncpg.pool.Pool) -> None:
+    LOGGER.debug("Connecting data classes")
+
+    for cls in _classes:
+        cls.set_connection_pool(pool)
+
+
+async def connect_pool(
+    host: str,
+    port: int,
+    database: str,
+    username: str,
+    password: str,
+    create_db_schema: bool = True,
+    connection_pool_min_size: int = 10,
+    connection_pool_max_size: int = 10,
+    connection_timeout: float = 60,
+) -> asyncpg.pool.Pool:
+
+    pool = await asyncpg.create_pool(
+        host=host,
+        port=port,
+        database=database,
+        user=username,
+        password=password,
+        min_size=connection_pool_min_size,
+        max_size=connection_pool_max_size,
+        timeout=connection_timeout,
+    )
+    try:
+        set_connection_pool(pool)
+        if create_db_schema:
+            async with pool.acquire() as con:
+                await schema.DBSchema(CORE_SCHEMA_NAME, PACKAGE_WITH_UPDATE_FILES, con).ensure_db_schema()
+            # expire connections after db schema migration to ensure cache consistency
+            await pool.expire_connections()
+        return pool
+    except Exception as e:
+        await pool.close()
+        await disconnect_pool()
+        raise e
+
+
+async def disconnect_pool() -> None:
+    LOGGER.debug("Disconnecting data classes")
+
+    # Enable `return_exceptions` to make sure we wait until all close_connection_pool() calls are finished
+
+    # or until the gather itself is cancelled.
+
+    result = await asyncio.gather(*[cls.close_connection_pool() for cls in _classes], return_exceptions=True)
+
+    exceptions = [r for r in result if r is not None and isinstance(r, Exception)]
+
+    if exceptions:
+        raise exceptions[0]
+
+
 async def start_engine(
     *,
     database_username: str,
     database_password: str,
     database_host: str,
     database_port: int,
     database_name: str,
-    pool_size: int = 10,
-    max_overflow: int = 0,
-    pool_timeout: float = 60.0,
+    pool: asyncpg.pool.Pool,
 ) -> None:
     """
```

Comment:
From what I understand, here we are creating our custom pool class so that sqlalchemy uses the asyncpg pool that we created previously. Is that correct?

Can you update the documentation for this part, mentioning what is  being done?

Reply:
I agree. This part needs a lot more documentation. Not only about what is done but also why.

Reply:
Updated the docstring.